src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.cs

```diff
@@ -16,6 +18,12 @@ namespace Microsoft.CodeAnalysis.PublicApiAnalyzers
     [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
     public sealed partial class DeclarePublicApiAnalyzer : DiagnosticAnalyzer
     {
+        /// <summary>
+        /// Cache from additional text instance to the api data we have read out for that specific file.  We only store
+        /// data for additional texts that explicitly match the public/internal api file names we expect.
+        /// </summary>
+        private static readonly ConditionalWeakTable<AdditionalText, ApiData> s_additionalTextToApiData = new();
```

Comment:
Please don’t use a CWT here. We already have an API on AnalysisContext to allow caching per-file data across compilations: https://github.com/dotnet/roslyn/blob/3a7a7407ea3c831630ecf2754092c33df3a6e452/src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs#L234

Reply:
Internally, we also use a CWT, so the API already provides a simple way to achieve caching:
[Note](https://github.com/dotnet/roslyn/blob/3a7a7407ea3c831630ecf2754092c33df3a6e452/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisValueProvider.cs#L13-L21)

Example test analyzer for this API: https://github.com/dotnet/roslyn/blob/3a7a7407ea3c831630ecf2754092c33df3a6e452/src/Compilers/Test/Core/Diagnostics/CommonDiagnosticAnalyzers.cs#L1605

Reply:
Sure.  I can do this.

Reply:
So i don't seem to be able to use that helper.  it is keyed off of SourceText.  But avoiding the source-text is the point here.  I need to keep off of AdditionalFile.  Thoughts @mavasani ?

Reply:
> But avoiding the source-text is the point here

I am not sure I understand. SourceText is already strongly held by the AdditionalText: https://github.com/dotnet/roslyn/blob/3a7a7407ea3c831630ecf2754092c33df3a6e452/src/Compilers/Core/Portable/AdditionalTextFile.cs#L49-L53

Isn't the key point here avoiding repeated compute of the data associated with each `AdditionalText` file?

Reply:
Ok.  Went deep down the rabit hole of trying to make this keyed off sourcetext.  It's painful.  Effectively, the data we compute off the source-text holds onto data outside of the source text (specifically, things like the path and other information about the original additionaltext).  I went down the path of trying to extract this out (e.g. RawApiData, ApiData, RawApiLine, ApiLine) and have the map point from the SourceText to the raw data.  But then there was so much wrapping of this i needed to do to make working with things palatable (and not have additional copies/allocatins happening).

If we make it so that hte context objects allow for storing arbitrary K/V and not just Text/Value or Tree/Value, then this will be trivial to do.

I discussed this with manish, and we decided that would take too much time.  So we're going to accept that this is a CWT.  However, i'm making it non-static so it has the same lifetime as the analyzer (And whatever owns it).